File: SEMANA 5.2/hash_tables_repo/python/hash_tables.py
"""
=============================================================================
Implementación de Tablas Hash en Python
=============================================================================
Práctica Avanzada - Estructuras de Datos (Semanas 5 y 6)
Universidad Nacional del Altiplano de Puno
Docente: Mg. Aldo Hernan Zanabria Galvez

Descripción:
    Implementa y compara tres estrategias de indexación:
      1. Tabla Hash con Encadenamiento Separado
      2. Tabla Hash con Sondeo Lineal
      3. Diccionario nativo de Python (referencia de rendimiento)

Dataset:
    eCommerce Behavior Data from Multi Category Store
    Autor: Mikhail Kechinov
    Fuente: https://www.kaggle.com/datasets/mkechinov/ecommerce-behavior-data-from-multi-category-store

Uso:
    1. Descarga el dataset de Kaggle y guárdalo como 'ecommerce_sample.csv'
       en la misma carpeta que este archivo.
    2. Ejecuta: python hash_tables.py

Dependencias:
    pip install pandas matplotlib
=============================================================================
"""

import pandas as pd
import time
import matplotlib.pyplot as plt
from typing import List, Tuple, Any, Optional


# =============================================================================
# CLASE 1: Tabla Hash con Encadenamiento Separado
# =============================================================================

class HashTableChaining:
    """
    Tabla hash que resuelve colisiones mediante encadenamiento separado.

    Cada posición del arreglo interno contiene una lista de pares (clave, valor).
    Cuando dos claves generan el mismo índice (colisión), el nuevo elemento
    se agrega al final de la lista correspondiente.

    Complejidad promedio:
        - Inserción: O(1)
        - Búsqueda:  O(1 + α)  donde α es el factor de carga
        - Eliminación: O(1 + α)
    """

    def __init__(self, size: int):
        """
        Inicializa la tabla hash.

        Args:
            size (int): Número de posiciones de la tabla.
                        Se recomienda usar un número primo para
                        mejorar la distribución de las claves.
        """
        self.size = size
        self.table: List[List[Tuple[Any, Any]]] = [[] for _ in range(size)]
        self.collisions = 0

    def hash_function(self, key) -> int:
        """
        Calcula el índice de una clave dentro de la tabla.

        Usa la función hash built-in de Python seguida de una operación
        módulo para obtener un índice válido.

        Args:
            key: La clave a transformar (cualquier tipo hashable).

        Returns:
            int: Índice entre 0 y self.size - 1.
        """
        return hash(str(key)) % self.size

    def insert(self, key, value) -> None:
        """
        Inserta un par (clave, valor) en la tabla hash.

        Si la posición calculada ya tiene elementos, se registra
        una colisión y el nuevo elemento se agrega a la lista.

        Args:
            key:   La clave de búsqueda.
            value: El valor asociado a la clave.
        """
        index = self.hash_function(key)

        # Si la lista en esa posición ya tiene elementos -> hay colisión
        if len(self.table[index]) > 0:
            self.collisions += 1

        self.table[index].append((key, value))

    def search(self, key):
        """
        Busca un valor por su clave.

        Calcula el índice de la clave y recorre la lista en esa
        posición hasta encontrar la clave exacta.

        Args:
            key: La clave a buscar.

        Returns:
            El valor asociado a la clave, o None si no existe.
        """
        index = self.hash_function(key)

        for k, v in self.table[index]:
            if k == key:
                return v

        return None  # Clave no encontrada

    def delete(self, key) -> bool:
        """
        Elimina un par (clave, valor) de la tabla.

        Args:
            key: La clave a eliminar.

        Returns:
            bool: True si se eliminó, False si la clave no existía.
        """
        index = self.hash_function(key)

        for i, (k, v) in enumerate(self.table[index]):
            if k == key:
                self.table[index].pop(i)
                return True

        return False

    def load_factor(self, total_elements: int) -> float:
        """
        Calcula el factor de carga α = n / m.

        Args:
            total_elements (int): Número de elementos insertados (n).

        Returns:
            float: Factor de carga entre 0.0 y 1.0 (o mayor si hay encadenamiento).
        """
        return total_elements / self.size

    def __repr__(self) -> str:
        return (f"HashTableChaining(size={self.size}, "
                f"collisions={self.collisions})")


# =============================================================================
# CLASE 2: Tabla Hash con Sondeo Lineal
# =============================================================================

class HashTableLinearProbing:
    """
    Tabla hash de direccionamiento abierto con sondeo lineal.

    Todos los elementos se almacenan directamente en el arreglo principal,
    sin estructuras auxiliares. Cuando una posición está ocupada, el
    algoritmo explora la siguiente posición disponible secuencialmente:

        h_i(k) = (h(k) + i) mod m,   i = 0, 1, 2, ...

    Complejidad promedio (con α < 0.5):
        - Inserción: O(1)
        - Búsqueda:  O(1 / (1 - α))
        - Eliminación: O(1 / (1 - α))

    Advertencia:
        El rendimiento se degrada significativamente cuando α > 0.7
        debido al fenómeno de agrupamiento primario.
    """

    def __init__(self, size: int):
        """
        Inicializa la tabla hash.

        Args:
            size (int): Capacidad máxima de la tabla. Debe ser mayor
                        al número de elementos que se insertarán.
        """
        self.size = size
        self.table: List[Optional[Tuple[Any, Any]]] = [None] * size
        self.collisions = 0

    def hash_function(self, key) -> int:
        """
        Calcula el índice inicial de una clave.

        Args:
            key: La clave a transformar.

        Returns:
            int: Índice entre 0 y self.size - 1.
        """
        return hash(str(key)) % self.size

    def insert(self, key, value) -> None:
        """
        Inserta un par (clave, valor) usando sondeo lineal.

        Si la posición inicial está ocupada, avanza secuencialmente
        hasta encontrar una posición libre. Cada intento fallido
        incrementa el contador de colisiones.

        Args:
            key:   La clave de búsqueda.
            value: El valor asociado.

        Raises:
            OverflowError: Si la tabla está completamente llena.
        """
        index = self.hash_function(key)
        start = index

        while self.table[index] is not None:
            self.collisions += 1
            index = (index + 1) % self.size

            # Si dimos la vuelta completa, la tabla está llena
            if index == start:
                raise OverflowError("La tabla hash está llena. "
                                    "Considera aumentar su tamaño.")

        self.table[index] = (key, value)

    def search(self, key):
        """
        Busca un valor usando sondeo lineal.

        Calcula el índice inicial y avanza hasta encontrar la clave
        o una posición vacía (lo que significa que la clave no existe).

        Args:
            key: La clave a buscar.

        Returns:
            El valor asociado a la clave, o None si no existe.
        """
        index = self.hash_function(key)
        start = index

        while self.table[index] is not None:
            if self.table[index][0] == key:
                return self.table[index][1]

            index = (index + 1) % self.size

            # Evitar bucle infinito
            if index == start:
                break

        return None

    def load_factor(self, total_elements: int) -> float:
        """
        Calcula el factor de carga α = n / m.

        Args:
            total_elements (int): Número de elementos insertados.

        Returns:
            float: Factor de carga. Debe mantenerse < 0.7 para
                   un rendimiento óptimo con sondeo lineal.
        """
        return total_elements / self.size

    def __repr__(self) -> str:
        return (f"HashTableLinearProbing(size={self.size}, "
                f"collisions={self.collisions})")


# =============================================================================
# FUNCIÓN: Graficar resultados comparativos
# =============================================================================

def plot_results(results: pd.DataFrame) -> None:
    """
    Genera dos gráficos de barras comparando los métodos evaluados.

    Args:
        results (pd.DataFrame): DataFrame con columnas:
            'Metodo', 'Tiempo insercion', 'Tiempo busqueda',
            'Colisiones', 'Factor de carga'.
    """
    fig, axes = plt.subplots(1, 2, figsize=(12, 5))

    # --- Gráfico 1: Tiempo de inserción ---
    axes[0].bar(results["Metodo"], results["Tiempo insercion"],
                color=["#2E5496", "#C55A11", "#538135"])
    axes[0].set_title("Comparación de tiempo de inserción", fontsize=13)
    axes[0].set_xlabel("Método")
    axes[0].set_ylabel("Tiempo (segundos)")
    axes[0].tick_params(axis='x', rotation=15)

    # --- Gráfico 2: Tiempo de búsqueda ---
    axes[1].bar(results["Metodo"], results["Tiempo busqueda"],
                color=["#2E5496", "#C55A11", "#538135"])
    axes[1].set_title("Comparación de tiempo de búsqueda", fontsize=13)
    axes[1].set_xlabel("Método")
    axes[1].set_ylabel("Tiempo (segundos)")
    axes[1].tick_params(axis='x', rotation=15)

    plt.tight_layout()
    plt.savefig("resultados_hash.png", dpi=150, bbox_inches="tight")
    plt.show()
    print("  -> Gráfico guardado como 'resultados_hash.png'")


# =============================================================================
# PROGRAMA PRINCIPAL
# =============================================================================

def main():
    print("=" * 60)
    print("  ANÁLISIS DE TABLAS HASH - DATOS KAGGLE")
    print("=" * 60)

    # ------------------------------------------------------------------
    # 1. Cargar y limpiar el dataset
    # ------------------------------------------------------------------
    print("\n[1] Cargando dataset...")
    import os
    
    # Buscar archivos CSV en el directorio Data (dos niveles arriba)
    data_dir = os.path.join(os.path.dirname(__file__), "..", "Data")
    
    csv_files = [f for f in os.listdir(data_dir) if f.endswith(".csv")]
    
    if not csv_files:
        print("\n  ERROR: No se encontraron archivos CSV en la carpeta Data.")
        return
    
    # Usar el primer archivo CSV encontrado
    csv_path = os.path.join(data_dir, csv_files[0])
    
    try:
        print(f"  -> Leyendo: {csv_files[0]}")
        df = pd.read_csv(csv_path, encoding='utf-8')
    except Exception as e:
        print(f"\n  ERROR: No se pudo leer el archivo: {e}\n")
        return

    df = df.dropna()
    print(f"  -> Filas cargadas (sin nulos): {len(df):,}")

    # ------------------------------------------------------------------
    # 2. Extraer claves (primeros 10,000 user_id)
    # ------------------------------------------------------------------
    keys = df["user_id"].astype(str).head(10_000).tolist()
    TABLE_SIZE = 20_011  # Número primo para mejor distribución

    print(f"  -> Claves extraídas: {len(keys):,}")
    print(f"  -> Tamaño de tabla:  {TABLE_SIZE:,}")

    # ------------------------------------------------------------------
    # 3. Prueba: Encadenamiento Separado
    # ------------------------------------------------------------------
    print("\n[2] Prueba con Encadenamiento Separado...")
    hash_chain = HashTableChaining(TABLE_SIZE)

    start = time.time()
    for key in keys:
        hash_chain.insert(key, {"user_id": key})
    insert_time_chain = time.time() - start

    start = time.time()
    for key in keys[:1_000]:
        hash_chain.search(key)
    search_time_chain = time.time() - start

    print(f"  -> Inserción: {insert_time_chain:.6f} s")
    print(f"  -> Búsqueda (1,000 claves): {search_time_chain:.6f} s")
    print(f"  -> Colisiones: {hash_chain.collisions:,}")
    print(f"  -> Factor de carga: {hash_chain.load_factor(len(keys)):.4f}")

    # ------------------------------------------------------------------
    # 4. Prueba: Sondeo Lineal
    # ------------------------------------------------------------------
    print("\n[3] Prueba con Sondeo Lineal...")
    hash_linear = HashTableLinearProbing(TABLE_SIZE)

    start = time.time()
    for key in keys:
        hash_linear.insert(key, {"user_id": key})
    insert_time_linear = time.time() - start

    start = time.time()
    for key in keys[:1_000]:
        hash_linear.search(key)
    search_time_linear = time.time() - start

    print(f"  -> Inserción: {insert_time_linear:.6f} s")
    print(f"  -> Búsqueda (1,000 claves): {search_time_linear:.6f} s")
    print(f"  -> Colisiones: {hash_linear.collisions:,}")
    print(f"  -> Factor de carga: {hash_linear.load_factor(len(keys)):.4f}")

    # ------------------------------------------------------------------
    # 5. Prueba: Diccionario nativo de Python (referencia)
    # ------------------------------------------------------------------
    print("\n[4] Prueba con Diccionario nativo de Python...")
    native_dict = {}

    start = time.time()
    for key in keys:
        native_dict[key] = {"user_id": key}
    insert_time_dict = time.time() - start

    start = time.time()
    for key in keys[:1_000]:
        native_dict.get(key)
    search_time_dict = time.time() - start

    print(f"  -> Inserción: {insert_time_dict:.6f} s")
    print(f"  -> Búsqueda (1,000 claves): {search_time_dict:.6f} s")
    print(f"  -> Colisiones: N/A (estructura nativa)")

    # ------------------------------------------------------------------
    # 6. Tabla comparativa de resultados
    # ------------------------------------------------------------------
    print("\n[5] Tabla comparativa de resultados:")
    results = pd.DataFrame({
        "Metodo": [
            "Encadenamiento",
            "Sondeo Lineal",
            "Diccionario Python"
        ],
        "Tiempo insercion": [
            insert_time_chain,
            insert_time_linear,
            insert_time_dict
        ],
        "Tiempo busqueda": [
            search_time_chain,
            search_time_linear,
            search_time_dict
        ],
        "Colisiones": [
            hash_chain.collisions,
            hash_linear.collisions,
            0
        ],
        "Factor de carga": [
            hash_chain.load_factor(len(keys)),
            hash_linear.load_factor(len(keys)),
            len(keys) / TABLE_SIZE
        ]
    })

    print()
    print(results.to_string(index=False))

    # ------------------------------------------------------------------
    # 7. Generar gráficos
    # ------------------------------------------------------------------
    print("\n[6] Generando gráficos...")
    plot_results(results)

    print("\n" + "=" * 60)
    print("  Ejecución completada exitosamente.")
    print("=" * 60)


if __name__ == "__main__":
    main()

File: SEMANA 5.2/hash_tables_repo/cpp/hash_tables.cpp
/*
=============================================================================
  Implementación de Tablas Hash en C++
=============================================================================
  Práctica Avanzada - Estructuras de Datos (Semanas 5 y 6)
  Universidad Nacional del Altiplano de Puno
  Docente: Mg. Aldo Hernan Zanabria Galvez

  Descripción:
    Implementa y compara dos estrategias de resolución de colisiones:
      1. Tabla Hash con Encadenamiento Separado (HashTableChaining)
      2. Tabla Hash con Sondeo Lineal (HashTableLinear)

  Dataset:
    eCommerce Behavior Data from Multi Category Store
    Autor: Mikhail Kechinov
    Fuente: https://www.kaggle.com/datasets/mkechinov/
            ecommerce-behavior-data-from-multi-category-store

  Compilación:
    g++ -std=c++17 -O2 -o hash_tables hash_tables.cpp

  Uso:
    ./hash_tables
    (El archivo 'ecommerce_sample.csv' debe estar en el mismo directorio)

  Dependencias:
    Biblioteca estándar de C++17 (no requiere instalaciones externas)
=============================================================================
*/

#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <list>
#include <chrono>
#include <string>
#include <iomanip>

using namespace std;
using namespace chrono;


// =============================================================================
// ESTRUCTURA: Registro clave-valor
// =============================================================================

/**
 * @brief Par clave-valor almacenado en la tabla hash.
 */
struct Record {
    string key;    ///< Clave de búsqueda (ej. user_id como string)
    string value;  ///< Valor asociado a la clave
};


// =============================================================================
// CLASE 1: Tabla Hash con Encadenamiento Separado
// =============================================================================

/**
 * @brief Tabla hash que resuelve colisiones mediante listas enlazadas.
 *
 * Cada posición del vector interno contiene una std::list<Record>.
 * Cuando dos claves generan el mismo índice (colisión), el nuevo
 * elemento se agrega al final de la lista correspondiente.
 *
 * Complejidad promedio:
 *   - Inserción: O(1)
 *   - Búsqueda:  O(1 + α)  donde α = n/m es el factor de carga
 */
class HashTableChaining {
private:
    int size;                          ///< Número de posiciones de la tabla
    vector<list<Record>> table;        ///< Vector de listas enlazadas
    int collisions;                    ///< Contador de colisiones registradas

public:

    /**
     * @brief Constructor.
     * @param tableSize Tamaño de la tabla. Se recomienda un número primo.
     */
    explicit HashTableChaining(int tableSize)
        : size(tableSize), collisions(0) {
        table.resize(size);
    }

    /**
     * @brief Función hash polinomial para cadenas.
     *
     * Calcula: hash = Σ(c * 31^i) mod m
     * El multiplicador 31 es un primo pequeño que minimiza colisiones
     * en distribuciones típicas de cadenas.
     *
     * @param  key Clave de tipo string.
     * @return Índice entre 0 y size - 1.
     */
    int hashFunction(const string& key) const {
        unsigned long hashVal = 0;
        for (char c : key) {
            hashVal = hashVal * 31 + static_cast<unsigned char>(c);
        }
        return static_cast<int>(hashVal % static_cast<unsigned long>(size));
    }

    /**
     * @brief Inserta un registro en la tabla hash.
     *
     * Si la posición ya contiene elementos se registra una colisión,
     * pero el nuevo elemento siempre se agrega a la lista.
     *
     * @param key   Clave del registro.
     * @param value Valor asociado.
     */
    void insert(const string& key, const string& value) {
        int index = hashFunction(key);

        if (!table[index].empty()) {
            collisions++;
        }

        table[index].push_back({key, value});
    }

    /**
     * @brief Busca un registro por su clave.
     *
     * @param  key Clave a buscar.
     * @return Valor asociado si se encontró, "No encontrado" en caso contrario.
     */
    string search(const string& key) const {
        int index = hashFunction(key);

        for (const Record& r : table[index]) {
            if (r.key == key) {
                return r.value;
            }
        }

        return "No encontrado";
    }

    /**
     * @brief Elimina un registro por su clave.
     *
     * @param  key Clave a eliminar.
     * @return true si se eliminó, false si no existía.
     */
    bool remove(const string& key) {
        int index = hashFunction(key);

        for (auto it = table[index].begin(); it != table[index].end(); ++it) {
            if (it->key == key) {
                table[index].erase(it);
                return true;
            }
        }

        return false;
    }

    /** @return Número total de colisiones registradas. */
    int getCollisions() const { return collisions; }

    /**
     * @brief Calcula el factor de carga α = n / m.
     * @param  totalElements Número de elementos insertados (n).
     * @return Factor de carga como valor decimal.
     */
    double loadFactor(int totalElements) const {
        return static_cast<double>(totalElements) / size;
    }
};


// =============================================================================
// CLASE 2: Tabla Hash con Sondeo Lineal
// =============================================================================

/**
 * @brief Tabla hash de direccionamiento abierto con sondeo lineal.
 *
 * Todos los elementos se almacenan directamente en el arreglo principal.
 * Cuando una posición está ocupada, el algoritmo busca la siguiente
 * posición libre de forma secuencial:
 *
 *   h_i(k) = (h(k) + i) mod m,   i = 0, 1, 2, ...
 *
 * Nota: El rendimiento se degrada cuando α > 0.7 por agrupamiento primario.
 */
class HashTableLinear {
private:
    int size;                    ///< Capacidad total de la tabla
    vector<Record> table;        ///< Arreglo principal de registros
    vector<bool> occupied;       ///< Marcas de ocupación por posición
    int collisions;              ///< Contador de intentos fallidos

public:

    /**
     * @brief Constructor.
     * @param tableSize Capacidad de la tabla. Debe superar al número
     *                  de elementos que se insertarán.
     */
    explicit HashTableLinear(int tableSize)
        : size(tableSize), collisions(0) {
        table.resize(size);
        occupied.assign(size, false);
    }

    /**
     * @brief Función hash polinomial (idéntica a HashTableChaining).
     *
     * @param  key Clave de tipo string.
     * @return Índice entre 0 y size - 1.
     */
    int hashFunction(const string& key) const {
        unsigned long hashVal = 0;
        for (char c : key) {
            hashVal = hashVal * 31 + static_cast<unsigned char>(c);
        }
        return static_cast<int>(hashVal % static_cast<unsigned long>(size));
    }

    /**
     * @brief Inserta un registro usando sondeo lineal.
     *
     * Si la posición inicial está ocupada, avanza secuencialmente
     * hasta encontrar una posición libre.
     *
     * @param key   Clave del registro.
     * @param value Valor asociado.
     */
    void insert(const string& key, const string& value) {
        int index = hashFunction(key);

        while (occupied[index]) {
            collisions++;
            index = (index + 1) % size;
        }

        table[index] = {key, value};
        occupied[index] = true;
    }

    /**
     * @brief Busca un registro por su clave usando sondeo lineal.
     *
     * @param  key Clave a buscar.
     * @return Valor asociado si se encontró, "No encontrado" en caso contrario.
     */
    string search(const string& key) const {
        int index = hashFunction(key);
        int start = index;

        while (occupied[index]) {
            if (table[index].key == key) {
                return table[index].value;
            }

            index = (index + 1) % size;

            // Evitar bucle infinito si la tabla está llena
            if (index == start) break;
        }

        return "No encontrado";
    }

    /** @return Número total de intentos fallidos de inserción. */
    int getCollisions() const { return collisions; }

    /**
     * @brief Calcula el factor de carga α = n / m.
     * @param  totalElements Número de elementos insertados.
     * @return Factor de carga como valor decimal.
     */
    double loadFactor(int totalElements) const {
        return static_cast<double>(totalElements) / size;
    }
};


// =============================================================================
// FUNCIÓN: Leer claves desde un archivo CSV
// =============================================================================

/**
 * @brief Lee las claves de la primera columna de un archivo CSV.
 *
 * Omite la fila de encabezado y lee hasta `limit` filas. Asume
 * que la primera columna contiene el user_id (o clave deseada).
 *
 * @param filename Ruta al archivo CSV.
 * @param limit    Número máximo de filas a leer.
 * @return Vector de strings con las claves extraídas.
 */
vector<string> readKeysFromCSV(const string& filename, int limit) {
    vector<string> keys;
    ifstream file(filename);

    if (!file.is_open()) {
        cerr << "\n  ERROR: No se pudo abrir '" << filename << "'.\n";
        cerr << "  Descarga el dataset desde:\n";
        cerr << "  https://www.kaggle.com/datasets/mkechinov/"
                "ecommerce-behavior-data-from-multi-category-store\n\n";
        return keys;
    }

    string line;
    getline(file, line);  // Omitir encabezado

    while (getline(file, line) && static_cast<int>(keys.size()) < limit) {
        stringstream ss(line);
        string column;
        vector<string> columns;

        while (getline(ss, column, ',')) {
            columns.push_back(column);
        }

        if (!columns.empty() && !columns[0].empty()) {
            keys.push_back(columns[0]);
        }
    }

    return keys;
}


// =============================================================================
// FUNCIÓN: Imprimir línea separadora
// =============================================================================

void printSeparator(char c = '-', int width = 60) {
    cout << string(width, c) << "\n";
}


// =============================================================================
// PROGRAMA PRINCIPAL
// =============================================================================

int main() {
    printSeparator('=');
    cout << "  ANÁLISIS DE TABLAS HASH - DATOS KAGGLE (C++)\n";
    printSeparator('=');

    // ------------------------------------------------------------------
    // Configuración del experimento
    // ------------------------------------------------------------------
    const string FILENAME   = "ecommerce_sample.csv";
    const int    LIMIT      = 10'000;
    const int    TABLE_SIZE = 20'011;  // Número primo
    const int    SEARCH_N   = 1'000;

    // ------------------------------------------------------------------
    // 1. Cargar claves desde el CSV
    // ------------------------------------------------------------------
    cout << "\n[1] Cargando dataset...\n";
    vector<string> keys = readKeysFromCSV(FILENAME, LIMIT);

    if (keys.empty()) return 1;

    cout << "  -> Claves cargadas: " << keys.size() << "\n";
    cout << "  -> Tamaño de tabla: " << TABLE_SIZE << "\n";

    // ------------------------------------------------------------------
    // 2. Prueba: Encadenamiento Separado
    // ------------------------------------------------------------------
    cout << "\n[2] Prueba con Encadenamiento Separado...\n";
    HashTableChaining hashChain(TABLE_SIZE);

    auto t0 = high_resolution_clock::now();
    for (const string& key : keys) {
        hashChain.insert(key, key);
    }
    auto t1 = high_resolution_clock::now();
    duration<double> insertChainTime = t1 - t0;

    t0 = high_resolution_clock::now();
    for (int i = 0; i < SEARCH_N && i < static_cast<int>(keys.size()); i++) {
        hashChain.search(keys[i]);
    }
    t1 = high_resolution_clock::now();
    duration<double> searchChainTime = t1 - t0;

    cout << fixed << setprecision(6);
    cout << "  -> Insercion:  " << insertChainTime.count() << " s\n";
    cout << "  -> Busqueda:   " << searchChainTime.count() << " s\n";
    cout << "  -> Colisiones: " << hashChain.getCollisions() << "\n";
    cout << "  -> Factor de carga: "
         << hashChain.loadFactor(static_cast<int>(keys.size())) << "\n";

    // ------------------------------------------------------------------
    // 3. Prueba: Sondeo Lineal
    // ------------------------------------------------------------------
    cout << "\n[3] Prueba con Sondeo Lineal...\n";
    HashTableLinear hashLinear(TABLE_SIZE);

    t0 = high_resolution_clock::now();
    for (const string& key : keys) {
        hashLinear.insert(key, key);
    }
    t1 = high_resolution_clock::now();
    duration<double> insertLinearTime = t1 - t0;

    t0 = high_resolution_clock::now();
    for (int i = 0; i < SEARCH_N && i < static_cast<int>(keys.size()); i++) {
        hashLinear.search(keys[i]);
    }
    t1 = high_resolution_clock::now();
    duration<double> searchLinearTime = t1 - t0;

    cout << "  -> Insercion:  " << insertLinearTime.count() << " s\n";
    cout << "  -> Busqueda:   " << searchLinearTime.count() << " s\n";
    cout << "  -> Colisiones: " << hashLinear.getCollisions() << "\n";
    cout << "  -> Factor de carga: "
         << hashLinear.loadFactor(static_cast<int>(keys.size())) << "\n";

    // ------------------------------------------------------------------
    // 4. Tabla resumen de resultados
    // ------------------------------------------------------------------
    cout << "\n[4] Resultados del experimento:\n";
    printSeparator();
    cout << left
         << setw(22) << "Metodo"
         << setw(16) << "Ins. (s)"
         << setw(16) << "Busq. (s)"
         << setw(14) << "Colisiones"
         << "Carga\n";
    printSeparator();
    cout << setw(22) << "Encadenamiento"
         << setw(16) << insertChainTime.count()
         << setw(16) << searchChainTime.count()
         << setw(14) << hashChain.getCollisions()
         << hashChain.loadFactor(static_cast<int>(keys.size())) << "\n";

    cout << setw(22) << "Sondeo Lineal"
         << setw(16) << insertLinearTime.count()
         << setw(16) << searchLinearTime.count()
         << setw(14) << hashLinear.getCollisions()
         << hashLinear.loadFactor(static_cast<int>(keys.size())) << "\n";
    printSeparator();

    cout << "\n";
    printSeparator('=');
    cout << "  Ejecucion completada exitosamente.\n";
    printSeparator('=');

    return 0;
}
